File: android-java/app/src/main/java/io/agora/convoai/example/voiceassistant/ui/AgentConfigFragment.java
package io.agora.convoai.example.voiceassistant.ui;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.fragment.NavHostFragment;
import io.agora.convoai.example.voiceassistant.KeyCenter;
import io.agora.convoai.example.voiceassistant.java.R;
import io.agora.convoai.example.voiceassistant.java.databinding.FragmentAgentConfigBinding;
import io.agora.convoai.example.voiceassistant.ui.common.BaseFragment;
import io.agora.convoai.example.voiceassistant.ui.common.SnackbarHelper;

import java.util.ArrayList;
import java.util.List;

/**
 * AgentConfigFragment - Configuration screen for starting a conversation
 */
public class AgentConfigFragment extends BaseFragment<FragmentAgentConfigBinding> {

    private ConversationViewModel viewModel;
    private boolean hasNavigated = false;
    private final List<String> statusHistory = new ArrayList<>();
    private String lastStatusMessage = "";

    @Override
    protected FragmentAgentConfigBinding getViewBinding(LayoutInflater inflater, ViewGroup container) {
        return FragmentAgentConfigBinding.inflate(inflater, container, false);
    }

    @Override
    public void initData() {
        super.initData();
        // Get ViewModel from activity scope
        viewModel = new ViewModelProvider(requireActivity()).get(ConversationViewModel.class);
    }

    @Override
    public void initView() {
        FragmentAgentConfigBinding binding = getBinding();
        if (binding == null) {
            return;
        }

        setOnApplyWindowInsets(binding.getRoot());

        // Display App ID first 5 characters
        String appId = KeyCenter.AGORA_APP_ID;
        String appIdPrefix = (appId != null && appId.length() >= 5) 
            ? appId.substring(0, 5) + "***" 
            : "*****";
        binding.tvAppId.setText(appIdPrefix);

        // Display Pipeline ID first 5 characters
        String pipelineId = KeyCenter.PIPELINE_ID;
        String pipelineIdPrefix = (pipelineId != null && pipelineId.length() >= 5) 
            ? pipelineId.substring(0, 5) + "***" 
            : "*****";
        binding.tvPipelineId.setText(pipelineIdPrefix);

        binding.btnStarter.setOnClickListener(v -> {
            // Generate random channel name each time joining channel
            String channelName = ConversationViewModel.generateRandomChannelName();

            // Clear status history when starting new connection
            statusHistory.clear();
            lastStatusMessage = "";
            updateStatusDisplay();

            // Check microphone permission before joining channel
            if (getActivity() instanceof MainActivity) {
                MainActivity mainActivity = (MainActivity) getActivity();
                mainActivity.checkMicrophonePermission(granted -> {
                    if (granted) {
                        viewModel.joinChannelAndLogin(channelName);
                    } else {
                        SnackbarHelper.showError(
                            AgentConfigFragment.this,
                            "Microphone permission is required to join channel"
                        );
                    }
                });
            }
        });

        // Observe UI state changes
        observeUiState();
    }

    private void observeUiState() {
        viewModel.uiState.observe(getViewLifecycleOwner(), state -> {
            FragmentAgentConfigBinding binding = getBinding();
            if (binding == null || state == null) {
                return;
            }

            // Reset hasNavigated flag when disconnected (after hangup)
            if (state.connectionState != ConversationViewModel.ConnectionState.Connected && hasNavigated) {
                hasNavigated = false;
            }

            // Update status - accumulate status messages for history
            updateStatusMessage(state);

            // Update loading state
            binding.btnStarter.setEnabled(state.connectionState != ConversationViewModel.ConnectionState.Connecting);
            binding.progressBar.setVisibility(
                state.connectionState == ConversationViewModel.ConnectionState.Connecting 
                    ? View.VISIBLE 
                    : View.GONE
            );

            // Show error via Snackbar if connection state is Error
            if (state.connectionState == ConversationViewModel.ConnectionState.Error &&
                state.statusMessage != null &&
                !state.statusMessage.isEmpty() &&
                isAdded() && isResumed()
            ) {
                SnackbarHelper.showError(this, state.statusMessage);
            }

            // Navigate to voice assistant when RTC and RTM are connected successfully (only once)
            if (state.connectionState == ConversationViewModel.ConnectionState.Connected && !hasNavigated) {
                hasNavigated = true;
                NavHostFragment.findNavController(this).navigate(R.id.action_agentConfig_to_voiceAssistant);
            }
        });
    }

    /**
     * Update status message - accumulate status messages for history
     * Only show when connecting, hide when idle/connected/disconnected
     */
    private void updateStatusMessage(ConversationViewModel.ConversationUiState state) {
        // Add new status message to history if it's different from the last one
        if (state.connectionState == ConversationViewModel.ConnectionState.Connecting &&
            state.statusMessage != null &&
            !state.statusMessage.isEmpty() &&
            !state.statusMessage.equals(lastStatusMessage)
        ) {
            statusHistory.add(state.statusMessage);
            lastStatusMessage = state.statusMessage;
        }

        // Clear history when disconnected or idle
        if (state.connectionState == ConversationViewModel.ConnectionState.Idle) {
            statusHistory.clear();
            lastStatusMessage = "";
        }

        updateStatusDisplay();
    }

    /**
     * Update status display with accumulated history
     */
    private void updateStatusDisplay() {
        FragmentAgentConfigBinding binding = getBinding();
        if (binding == null) {
            return;
        }

        StringBuilder displayMessage = new StringBuilder();
        for (int i = 0; i < statusHistory.size(); i++) {
            if (i > 0) {
                displayMessage.append("\n");
            }
            displayMessage.append(statusHistory.get(i));
        }
        binding.tvStatus.setText(displayMessage.toString());
    }
}

File: android-kotlin/app/src/main/java/io/agora/convoai/example/voiceassistant/ui/AgentConfigFragment.kt
package io.agora.convoai.example.voiceassistant.ui

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.activityViewModels
import io.agora.convoai.example.voiceassistant.ui.common.SnackbarHelper
import io.agora.convoai.example.voiceassistant.R
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.findNavController
import io.agora.convoai.example.voiceassistant.KeyCenter
import io.agora.convoai.example.voiceassistant.databinding.FragmentAgentConfigBinding
import io.agora.convoai.example.voiceassistant.ui.common.BaseFragment
import kotlinx.coroutines.launch

class AgentConfigFragment : BaseFragment<FragmentAgentConfigBinding>() {

    private val viewModel: ConversationViewModel by activityViewModels()
    private var hasNavigated = false
    private val statusHistory = mutableListOf<String>()
    private var lastStatusMessage = ""

    override fun getViewBinding(inflater: LayoutInflater, container: ViewGroup?): FragmentAgentConfigBinding? {
        return FragmentAgentConfigBinding.inflate(inflater, container, false)
    }

    override fun initView() {
        mBinding?.apply {
            setOnApplyWindowInsets(root)

            // Display App ID first 5 characters
            val appIdPrefix = KeyCenter.AGORA_APP_ID.take(5) + "***"
            tvAppId.text = appIdPrefix

            // Display Pipeline ID first 5 characters
            val pipelineIdPrefix = KeyCenter.PIPELINE_ID.take(5) + "***"
            tvPipelineId.text = pipelineIdPrefix

            btnStarter.setOnClickListener {
                // Generate random channel name each time joining channel
                val channelName = ConversationViewModel.generateRandomChannelName()

                // Clear status history when starting new connection
                statusHistory.clear()
                lastStatusMessage = ""
                updateStatusDisplay()

                // Check microphone permission before joining channel
                val mainActivity = activity as? MainActivity ?: return@setOnClickListener
                mainActivity.checkMicrophonePermission { granted ->
                    if (granted) {
                        viewModel.joinChannelAndLogin(channelName)
                    } else {
                        SnackbarHelper.showError(
                            this@AgentConfigFragment,
                            "Microphone permission is required to join channel"
                        )
                    }
                }
            }
        }

        // Observe UI state changes
        observeUiState()
    }

    private fun observeUiState() {
        viewLifecycleOwner.lifecycleScope.launch {
            viewModel.uiState.collect { state ->
                mBinding?.apply {
                    // Reset hasNavigated flag when disconnected (after hangup)
                    if (state.connectionState != ConversationViewModel.ConnectionState.Connected && hasNavigated) {
                        hasNavigated = false
                    }

                    // Update status - accumulate status messages for history
                    updateStatusMessage(state)

                    // Update loading state
                    btnStarter.isEnabled = state.connectionState != ConversationViewModel.ConnectionState.Connecting
                    progressBar.visibility =
                        if (state.connectionState == ConversationViewModel.ConnectionState.Connecting) View.VISIBLE else View.GONE

                    // Show error via Snackbar if connection state is Error
                    if (state.connectionState == ConversationViewModel.ConnectionState.Error &&
                        state.statusMessage.isNotEmpty() &&
                        isAdded && isResumed
                    ) {
                        SnackbarHelper.showError(this@AgentConfigFragment, state.statusMessage)
                    }

                    // Navigate to voice assistant when RTC and RTM are connected successfully (only once)
                    if (state.connectionState == ConversationViewModel.ConnectionState.Connected && !hasNavigated) {
                        hasNavigated = true
                        findNavController().navigate(R.id.action_agentConfig_to_voiceAssistant)
                    }
                }
            }
        }
    }

    /**
     * Update status message - accumulate status messages for history
     * Only show when connecting, hide when idle/connected/disconnected
     */
    private fun updateStatusMessage(state: ConversationViewModel.ConversationUiState) {
        // Add new status message to history if it's different from the last one
        if (state.connectionState == ConversationViewModel.ConnectionState.Connecting &&
            state.statusMessage.isNotEmpty() &&
            state.statusMessage != lastStatusMessage
        ) {
            statusHistory.add(state.statusMessage)
            lastStatusMessage = state.statusMessage
        }

        // Clear history when disconnected or idle
        if (state.connectionState == ConversationViewModel.ConnectionState.Idle) {
            statusHistory.clear()
            lastStatusMessage = ""
        }

        updateStatusDisplay()
    }

    /**
     * Update status display with accumulated history
     */
    private fun updateStatusDisplay() {
        mBinding?.apply {
            val displayMessage = statusHistory.joinToString("\n")
            tvStatus.text = displayMessage
        }
    }
}

